### PR TITLE
Create UC external locations in Azure based on migrated storage credentials

### LINKED ISSUE #100: Create EXTERNAL LOCATIONs to map to External Tables (Azure)
Related issues:
- https://github.com/databrickslabs/ucx/issues/339
- https://github.com/databrickslabs/ucx/issues/862
- https://github.com/databrickslabs/ucx/issues/338

Jobs to be done:

- use `ExternalLocations` from https://github.com/databrickslabs/ucx/blob/main/src/databricks/labs/ucx/hive_metastore/locations.py and `tables.TablesCrawler` to verify if we have matching external locations configured already - use `ws.external_locations.list()` API for this.
- validate the logic of `ExternalLocations` for correctness and corner-cases
- create `databricks labs ucx validate-external-locations` command to display a number of tables per external location and any HMS table locations that don't belong to an existing external location
- print Terraform resource definitions to have those external locations setup

### PR DESCRIPTION
**Pull request by qziyuan**
## Changes

Create UC external locations for missing locations reported by `ExternalLocations` in `hive_metastore/locations.py`.

### Linked issues

Partially resolve #100 

### Functionality 

- [ ] added relevant user documentation
- [x] added new CLI command
- [ ] modified existing command: `databricks labs ucx ...`
- [ ] added a new workflow
- [ ] modified existing workflow: `...`
- [ ] added a new table
- [ ] modified existing table: `...`

### Tests

- [ ] manually tested
- [x] added unit tests
- [x] added integration tests
- [ ] verified on staging environment (screenshot attached)

### CODE REVIEW

**Review by nfx — CHANGES_REQUESTED** [review]
logic seems okay at a glance, cover it with tests - it's hard to review the code without them:
image

File: src/databricks/labs/ucx/azure/locations.py
Comment (nfx) [inline src/databricks/labs/ucx/azure/locations.py:60]: Nit: This code can be de-nested with if-not-continue

File: src/databricks/labs/ucx/azure/locations.py
Comment (nfx) [inline src/databricks/labs/ucx/azure/locations.py:93]: Oh, nice - didn't know about this method in python

File: src/databricks/labs/ucx/azure/locations.py
Comment (nfx) [inline src/databricks/labs/ucx/azure/locations.py:85]: This data structure is incorrect, because if prefix has multiple readers or writers - the last one overrides the previous.

Correct data structure is dict[str,set[str]]

File: src/databricks/labs/ucx/azure/locations.py
Comment (nfx) [inline src/databricks/labs/ucx/azure/locations.py:53]: Nit: here the code would be more readable if you extract this as a variable. Same for the next condition

File: src/databricks/labs/ucx/azure/locations.py
Comment (nfx) [inline src/databricks/labs/ucx/azure/locations.py:57]: This statement is redundant

File: src/databricks/labs/ucx/azure/locations.py
Comment (nfx) [inline src/databricks/labs/ucx/azure/locations.py:105]: This is loosing data - what if the prefix contains a path?

File: src/databricks/labs/ucx/azure/locations.py
Comment (nfx) [inline src/databricks/labs/ucx/azure/locations.py:143]: What is [1] there?

File: src/databricks/labs/ucx/azure/locations.py
Comment (nfx) [inline src/databricks/labs/ucx/azure/locations.py:149]: Missing handling for external locations that are already there - what if this command runs twice? What if locations are created by terraform or by hand? What if only part of them are created by hand?

File: src/databricks/labs/ucx/azure/locations.py
Comment (nfx) [inline src/databricks/labs/ucx/azure/locations.py:168]: Else is redundant

File: src/databricks/labs/ucx/azure/resources.py
Comment (nfx) [inline src/databricks/labs/ucx/azure/resources.py:266]: Nit: extract a var to make more readable

File: src/databricks/labs/ucx/hive_metastore/locations.py
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/locations.py:163]: Return type is incorrect. Add more unit tests for this method if you make it public

File: tests/integration/azure/test_locations.py
Comment (nfx) [inline tests/integration/azure/test_locations.py:25]: Okay, as a follow-up after this PR is merged, make a PR to labs-infra which will setup these with Terraform. Or maybe within this PR. I'll do you a quick intro

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/azure/locations.py
Comment (qziyuan) [reply src/databricks/labs/ucx/azure/locations.py:85]: wWell this is intentional. We cannot have multiple UC external locations pointing to same prefix, so when multiple credentials have access to one same prefix, we can only pick one credential to create the external location. So the logic is first try pick one that has write access, if not pick one that has read access. So the data structure does not need to contain all readers and writers

File: src/databricks/labs/ucx/azure/locations.py
Comment (qziyuan) [reply src/databricks/labs/ucx/azure/locations.py:105]: `access.py` crawls [container to principal mapping](https://github.com/databrickslabs/ucx/blob/e0da29aee2c10e91f01b1c315812fa0c3e3eadf9/src/databricks/labs/ucx/azure/access.py#L79), so for a given location we extract the container url and use it as the key to find the principal and hence the storage credential to be used.

File: tests/integration/azure/test_locations.py
Comment (qziyuan) [reply tests/integration/azure/test_locations.py:25]: The storage account, managed identities etc. I'm using here belong to the field-eng subscription and they are subject to be removed with `removeAfter` tag.
If we use Terraform to populate them in labs-infra, the problem is I don't have read permission for Azure resources in labs-infra, those azure graph and resource manager api call to get application_id will fail.

File: src/databricks/labs/ucx/azure/locations.py
Comment (qziyuan) [reply src/databricks/labs/ucx/azure/locations.py:149]: The `match_table_external_locations()` will first call UC apt to list existed UC external locations. So the returned `missing_locs` will not include any created UC external locations no matter how many times this command runs.

File: src/databricks/labs/ucx/azure/resources.py
Comment (qziyuan) [reply src/databricks/labs/ucx/azure/resources.py]: fixed.

File: src/databricks/labs/ucx/azure/locations.py
Comment (qziyuan) [reply src/databricks/labs/ucx/azure/locations.py]: fixed

File: src/databricks/labs/ucx/azure/locations.py
Comment (qziyuan) [reply src/databricks/labs/ucx/azure/locations.py]: fixed.

File: src/databricks/labs/ucx/azure/locations.py
Comment (qziyuan) [reply src/databricks/labs/ucx/azure/locations.py]: fixed.

File: src/databricks/labs/ucx/azure/locations.py
Comment (qziyuan) [reply src/databricks/labs/ucx/azure/locations.py]: fixed.

File: src/databricks/labs/ucx/azure/locations.py
Comment (qziyuan) [reply src/databricks/labs/ucx/azure/locations.py]: refactored for better readability

**Review by nfx — APPROVED** [review]
Lgtm

**Pull request merged**
